src/relax/backend/vm/rewrite_cuda_graph.cc

```diff
@@ -0,0 +1,433 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include <tvm/relax/backend.h>
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/type.h>
+#include <tvm/tir/op.h>
+
+namespace tvm {
+namespace relax {
+
+/*!
+ * \brief Collect bindings that are the output of the funciton.
+ */
+class OutputCollector : public ExprVisitor {
+ public:
+  static std::unordered_set<const VarNode*> Collect(const Function& func) {
+    OutputCollector collector;
+    collector.VisitExpr(func->body);
+    return collector.outputs_;
+  }
+
+  void VisitExpr_(const VarNode* var) final {
+    if (is_visiting_output_) {
+      outputs_.insert(var);
+    }
+  }
+
+  void VisitExpr_(const SeqExprNode* seq) final {
+    bool is_visiting_output = false;
+    std::swap(is_visiting_output, is_visiting_output_);
+    for (const auto& binding_block : seq->blocks) {
+      VisitBindingBlock(binding_block);
+    }
+    std::swap(is_visiting_output, is_visiting_output_);
+    VisitExpr(seq->body);
+  }
+
+ private:
+  std::unordered_set<const VarNode*> outputs_;
+  bool is_visiting_output_ = true;
+};
+
+/*! \brief A statically executed region of the graph. A statically executed region is a region where
+ * all the tensors are statically allocated and only contains kernel launches (control flow is not
+ * allowed).
+ */
+struct StaticRegion {
```

Comment:
NOTE: this name is too generic and can easily cause name conflict -- sometimes causes silent bug. Consider put it under an anoymous namespace or a longer name